packages/playground/blueprints/src/lib/utils/logger.ts

```diff
@@ -0,0 +1,140 @@
+    Info,
+    // A warning event. Not an error but is likely more important than an informational event.
+    Warn,
+    // An error event. Something went wrong.
+    Error,
+    // A fatal error such as application or system crash.
+    Fatal,
+}
+
+/**
+ * A logger for Playground.
+ */
+export class Logger {
+    private readonly LOG_PREFIX = 'Playground';
+
+    /**
+     * The length of the last PHP log.
+     */
+    private lastPHPLogLength = 0;
+
+    constructor() {
+        this.collectPlaygroundLogs();
+    }
+
+    /**
+     * Collect errors from Playground and log them.
+     */
+    private collectPlaygroundLogs() {
+         if (typeof window !== 'undefined') {
+            window.addEventListener('error', (event) => {
+                this.log(
+                    `${event.message} in ${event.filename} on line ${event.lineno}:${event.colno}`,
+                    LogSeverity.Fatal
+                );
+            });
+            window.addEventListener('unhandledrejection', (event) => {
+                this.log(
+                    `${event.reason.stack}`,
+                    LogSeverity.Fatal
+                );
+            });
+            window.addEventListener('rejectionhandled', (event) => {
+                this.log(
+                    `${event.reason.stack}`,
+                    LogSeverity.Fatal
+                );
+            });
+        }
+    }
+
+    /**
+     * Register a listener for the request.end event and log the data.
+     * @param UniversalPHP playground instance
+     */
+    public addPlaygroundRequestEndListener(playground: UniversalPHP) {
+        playground.addEventListener('request.end', (event) => {
+            event = event as PHPRequestEndEvent;
+            if (event.data && event.data['log']) {
+                this.logRaw(event?.data['log'].substring(this.lastPHPLogLength));
+                this.lastPHPLogLength = event.data['log'].length;
+            }
+        } );
+    }
+
+
+    /**
+     * Get UTC date in the PHP log format https://github.com/php/php-src/blob/master/main/main.c#L849
+     *
+     * @param date
+     * @returns string
+     */
+    private formatLogDate(date: Date): string {
+        const day = String(date.getUTCDate()).padStart(2, '0');
+        const monthNames = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
+        const month = monthNames[date.getUTCMonth()];
+        const year = date.getUTCFullYear();
+        const hours = String(date.getUTCHours()).padStart(2, '0');
+        const minutes = String(date.getUTCMinutes()).padStart(2, '0');
+        const seconds = String(date.getUTCSeconds()).padStart(2, '0');
+
+        return `${day}-${month}-${year} ${hours}:${minutes}:${seconds} UTC`;
+    }
+
+    /**
+     * Format log message and severity and log it.
+     * @param string message
+     * @param LogSeverity severity
+     */
+    public formatLog(message: string, severity: LogSeverity): string {
```

Comment:
My first thought was "oh, is this logging a formatted message?" Maybe let's call it `formatMessage` to avoid confusion?

Reply:
Done [cabbd9b](https://github.com/WordPress/wordpress-playground/pull/1035/commits/cabbd9b66a996f7cf60384ea23be10a43e95768a)